app/src/main/java/org/thoughtcrime/securesms/conversation/v2/mention/MentionViewModel.kt

```diff
@@ -0,0 +1,272 @@
+    groupDatabase: GroupDatabase,
+    mmsDatabase: MmsDatabase,
+    contactDatabase: SessionContactDatabase,
+    memberDatabase: GroupMemberDatabase,
+    storage: Storage,
+) : ViewModel() {
+    private val editable = MentionEditable()
+
+    /**
+     * A factory that creates a new [Editable] instance that is backed by the same source of truth
+     * used by this viewModel.
+     */
+    val editableFactory = object : Editable.Factory() {
+        override fun newEditable(source: CharSequence?): Editable {
+            if (source === editable) {
+                return source
+            }
+
+            if (source != null) {
+                editable.replace(0, editable.length, source)
+            }
+
+            return editable
+        }
+    }
+
+    @Suppress("OPT_IN_USAGE")
+    private val members: StateFlow<List<Member>?> =
+        (contentResolver.observeChanges(Conversation.getUriForThread(threadID)) as Flow<Any?>)
+            .debounce(500L)
+            .onStart { emit(Unit) }
+            .mapLatest {
+                val r = checkNotNull(threadDatabase.getRecipientForThreadId(threadID)) {
+                    "Recipient not found for thread ID: $threadID"
+                }
+
+                val memberIDs = when {
+                    r.isClosedGroupRecipient -> {
+                        groupDatabase.getGroupMemberAddresses(r.address.toGroupString(), false)
+                            .map { it.serialize() }
+                    }
+
+                    r.isCommunityRecipient -> mmsDatabase.getRecentChatMemberIDs(threadID, 20)
+                    r.isContactRecipient -> listOf(r.address.serialize())
+                    else -> listOf()
+                }
+
+                val moderatorIDs = if (r.isCommunityRecipient) {
+                    val groupId = storage.getOpenGroup(threadID)?.id
+                    if (groupId.isNullOrBlank()) {
+                        emptySet()
+                    } else {
+                        memberDatabase.getGroupMembersRoles(groupId, memberIDs)
+                            .mapNotNullTo(hashSetOf()) { (memberId, roles) ->
+                                memberId.takeIf { roles.any { it.isModerator } }
+                            }
+                    }
+                } else {
+                    emptySet()
+                }
+
+                val contactContext = if (r.isCommunityRecipient) {
+                    Contact.ContactContext.OPEN_GROUP
+                } else {
+                    Contact.ContactContext.REGULAR
+                }
+
+                contactDatabase.getContacts(memberIDs).map { contact ->
+                    Member(
+                        publicKey = contact.sessionID,
+                        name = contact.displayName(contactContext).orEmpty(),
+                        isModerator = contact.sessionID in moderatorIDs,
+                    )
+                }
+            }
+            .flowOn(Dispatchers.IO)
```

Comment:
Are we ok with hardcoding dispatchers? Shouldn't we have an injectable interface to allow us to swap these when testing viewmodel functions?

Reply:
My understanding is that we don't need to swap out the dispatcher for testing, there's already test code for this class. But keen to hear if there's any other advantage of swapping out though.

Reply:
There normally are issues with hardcoded dispatchers. We normally swap them with  TestDispatchers in tests in order to assure deterministic results.

Reply:
Alright currently there's no deterministic issue with the IO dispatcher but happy to use a DI approach for this.